components/message.rst

```diff
@@ -0,0 +1,284 @@
+
+Same bus received and sender
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+To allow us to receive and send messages on the same bus and prevent a loop, the
+message bus is equipped with the ``WrapIntoReceivedMessage`` received. It will
+wrap the received messages into ``ReceivedMessage`` objects and the
+``SendMessageMiddleware`` middleware will know it should not send these messages.
+
+Your own sender
+---------------
+
+Using the ``SenderInterface``, you can easily create your own message sender.
+Let's say you already have an ``ImportantAction`` message going through the
+message bus and handled by a handler. Now, you also want to send this message as
+an email.
+
+First, create your sender::
+
+    namespace App\MessageSender;
+
+    use Symfony\Component\Message\SenderInterface;
+    use App\Message\ImportantAction;
+
+    class ImportantActionToEmailSender implements SenderInterface
+    {
+       private $toEmail;
+       private $mailer;
+
+       public function __construct(\Swift_Mailer $mailer, string $toEmail)
+       {
+           $this->mailer = $mailer;
+           $this->toEmail = $toEmail;
+       }
+
+       public function send($message)
+       {
+           if (!$message instanceof ImportantAction) {
+               throw new \InvalidArgumentException(sprintf('Producer only supports "%s" messages.', ImportantAction::class));
+           }
+
+           $this->mailer->send(
+               (new \Swift_Message('Important action made'))
+                   ->setTo($this->toEmail)
+                   ->setBody(
+                       '<h1>Important action</h1><p>Made by '.$message->getUsername().'</p>',
+                       'text/html'
+                   )
+           );
+       }
+    }
+
+Then, register your sender service:
+
+.. code-block:: yaml
+
+    services:
+         App\MessageSender\ImportantActionToEmailSender:
+             arguments:
+                 - "@mailer"
+                 - "%to_email%"
+
+             tags:
+                 - message.sender
+
+Finally, route your important message to the sender:
+
+.. code-block:: yaml
+
+    framework:
+       message:
+           routing:
+               'App\Message\ImportantAction': [App\MessageSender\ImportantActionToEmailSender, ~]
+
+.. note::
+
+    This example shows you how you can at the same time send your message and
+    directly handle it using a ``null`` (``~``) sender.
+
+Your own receiver
+-----------------
+
+A consumer is responsible for receiving messages from a source and dispatching
+them to the application.
+
+Let's say you already proceed some "orders" on your application using a
+``NewOrder`` message. Now you want to integrate with a 3rd party or a legacy
+application but you can't use an API and need to use a shared CSV file with new
+orders.
+
+You will read this CSV file and dispatch a ``NewOrder`` message. All you need to
+do is your custom CSV consumer and Symfony will do the rest.
```

Comment:
"to do" => "to do is create your…" or "to write is your…

Reply:
consumer -> receiver?